modules/javafx.base/src/main/java/javafx/beans/Subscription.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package javafx.beans;
+
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * A subscription encapsulates how to cancel it without having
+ * to keep track of how it was created.
```

Comment:
I'm not entirely sure about this; the `Subscription` interface is pretty neutral in how it can be used, and does not have any direct link to observables at all; it's more of a token that you get that can be used to cancel a previous action, it doesn't actively do anything (unless you count the wrapper that it puts around the provided function).

As it is, they don't even need to be associated with `Observable` or `ObservableValue` at all.  Just as easily you could do:

       FileInputStream inputStream = .. ;
       Subscription combined = Subscription.combine(
              property.subscribe(this::invalidated),
              () -> inputStream.close()
       );

       combined.subscribe();  // closes input stream and removes invalidations subscriber

Although I think it is a good description, I'm not sure it belongs on this class.

Reply:
I put the your comments (that we agree upon so far) in #1177 

As I said I'm not sure about the description for the Subscription class, and it's not included yet, and so I also haven't yet removed the duplicated texts about GC-ability from the `subscribe` methods.

Reply:
I also now wonder if `Subscription` perhaps is in the wrong package (javafx.beans currently).  Perhaps it should have been in javafx.util ?

Reply:
Good point, I didn't think about subscriptions outside of the listeners contexts. let alone outside of the JavaFX content. It does look now like it belongs in the utils package. Maybe a quick correction can be made if Kevin agrees with this and you think it's worth it.

Can you think where else in JavaFX subscriptions could be used? I'm thinking about the deprecated `finalize` and try-with-resource, anywhere where we need to release resources. This would give a better idea of what the class revolves around and thus what documentation fits it.

Maybe the class should have a section about its use with JavaFX observables. I'm looking at the [`Service`](https://openjfx.io/javadoc/20/javafx.graphics/javafx/concurrent/Service.html) and [`Taks`](https://openjfx.io/javadoc/20/javafx.graphics/javafx/concurrent/Task.html) classes, which can also be used as general purpose, but have clear uses within JavaFX.
So something like this structure:

```
[General info about the class]
How it can be used to release resources or finalize or undo... (the unsubscribe method)
<p>
Info about its `combine` and `and` methods (my 3rd paragraph)

# In JavaFX Observables [Specific info about its use with observables]
My paragraphs 1, 2, 4, 5

#Maybe some other wide specific use
```

Reply:
I also hadn't considered potential usage of Subscription outside of observables, but I can see some uses for it.

John: I would support moving it to javafx.util (via a separate issue) if you think it is worth doing.

Reply:
Correct packaging is often overlooked, so I think it is worth doing.  I'll test to see if it doesn't cause any problems, and if not, I'll create an issue for it.

Reply:
> Good point, I didn't think about subscriptions outside of the listeners contexts, let alone outside of the JavaFX content. It does look now like it belongs in the utils package. Maybe a quick correction can be made if Kevin agrees with this and you think it's worth it.
> 
> Can you think where else in JavaFX subscriptions could be used? I'm thinking about the deprecated `finalize` and try-with-resource, anywhere where we need to release resources. This would give a better idea of what the class revolves around and thus what documentation fits it.

It's very general, basically a specifically purposed `Runnable` where `run` is called `unsubscribe` with some chaining methods. You could use it for purposes we may not be able to think of right now.  I've looked at some `dispose` methods, which often have clean-up code for some inspiration:

1. Use it to clean up references: `() -> { this.skinnable = null; this.scrollBar = null; }`
2. Use it to stop animation timers or timelines : `() -> animationTimer.stop();`
3. You could wrap other add/remove style API that don't offer a `Subscription`: `() -> getSkinnable().removeEventHandler( ... )`

I'm not saying you always should do this, but if you are releasing a `Subscription` anyway, you could combine some of these.  Also, if new API decides to provide a subscription, they're not limited to only remove a listener, they could also stop a thread, a timer or release some resources that was started/allocated when the subscription was created.